File: src/com/demonwav/mcdev/platform/ProjectConfiguration.java
package com.demonwav.mcdev.platform;

import com.demonwav.mcdev.buildsystem.BuildSystem;

import com.intellij.openapi.module.Module;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public abstract class ProjectConfiguration {

    public String pluginName = null;
    public String pluginVersion = null;
    public String mainClass = null;
    public String description = null;
    public final List<String> authors = new ArrayList<>();

    public abstract void create(@NotNull Module module, @NotNull PlatformType type, @NotNull BuildSystem buildSystem);

    public boolean hasAuthors() {
        return listContainsAtLeastOne(this.authors);
    }

    public void setAuthors(String string) {
        this.authors.clear();
        Collections.addAll(this.authors, commaSplit(string));
    }

    public boolean hasDescription() {
        return description != null && !description.trim().isEmpty();
    }

    protected static String[] commaSplit(String string) {
        return string.trim().replaceAll("\\[|\\]", "").split("\\s*,\\s*");
    }

    protected static boolean listContainsAtLeastOne(List<String> list) {
        if (list == null || list.size() == 0) {
            return false;
        }

        Iterator<String> it = list.iterator();
        while (it.hasNext()) {
            String s = it.next();
            if (s.trim().isEmpty()) {
                it.remove();
            }
        }

        return list.size() != 0;
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/ProjectConfiguration.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2018 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform

import com.demonwav.mcdev.buildsystem.BuildSystem
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.util.containers.isNullOrEmpty

private val bracketRegex = Regex("[\\[\\]]")
private val commaRegex = Regex("\\s*,\\s*")

abstract class ProjectConfiguration {

    data class BaseConfigs(
        val pluginName: String,
        val pluginVersion: String,
        val mainClass: String,
        val description: String? = null,
        val website: String? = null,
        val authors: MutableList<String> = mutableListOf()
    )

    var base: BaseConfigs? = null

    abstract var type: PlatformType

    abstract fun create(project: Project, buildSystem: BuildSystem, indicator: ProgressIndicator)

    fun hasAuthors() = listContainsAtLeastOne(base?.authors)
    fun setAuthors(string: String) {
        base?.authors?.let { authors ->
            authors.clear()
            authors.addAll(commaSplit(string))
        }
    }

    fun hasDescription() = base?.description?.isNotBlank() == true

    protected fun commaSplit(string: String): List<String> {
        return if (!string.isBlank()) {
            string.trim().replace(bracketRegex, "").split(commaRegex).toList()
        } else {
            emptyList()
        }
    }

    fun listContainsAtLeastOne(list: MutableList<String>?): Boolean {
        if (list.isNullOrEmpty()) {
            return false
        }

        list?.removeIf(String::isBlank)

        return list?.size != 0
    }

    protected fun getMainClassDirectory(files: Array<String>, file: VirtualFile): VirtualFile {
        var movingFile = file
        for (i in 0 until (files.size - 1)) {
            val s = files[i]
            val temp = movingFile.findChild(s)
            movingFile = if (temp != null && temp.isDirectory) {
                temp
            } else {
                movingFile.createChildDirectory(this, s)
            }
        }

        return movingFile
    }
}
